### DISCUSSION TITLE
How to run agent in a subprocess started with Flask ?

### DISCUSSION THREAD

**Discussion by tsonglew** [apache/skywalking#10811] [Q&A] [answered]
Code :

```python
from skywalking import agent, config
config.init()
agent.start()

from flask import Flask

app = Flask(__name__)

@app.route('/hello/')
def hello_world():
    return 'Hello World!'


if __name__ == '__main__':
    app.run(processes=2, threaded=False)
```

When running flask application with multiple processes, the python agent only started in main process. It seems that Flask uses werkzeug by default, and does not provided any methods like `post_fork` in gunicorn. How can I start a agent in subprocess ?

**Comment by Superskyyy** [comment, marked as answer]
@tsonglew Hi! The Werkzeug's development server you are using is not safe for production usage so we don't consider it a normal use of the agent(and you should not use it in production). 

Anyway if it uses fork under the hood it can theoretically be automatically working with the prefork flag in the SkyWalking Python agent non-intrusive CLI. But if I recall correctly it's not enabled now, it will require some changes to the CLI code (we currently only allow uwsgi and gunicorn to be automatically handled upon fork, there is a if check for command starts with uwsgi/gunicorn. )

Since you are using app.run instead of a CLI run another way is to change the code in agent/init logic to expose a new api that forcefully recreate all threads to the agent so they can work in the child processes. Since preforking models are generally less popular these days, we are moving to embrace ASGI/Asyncio, these forking related convenience features may not be added in upstream anymore.
